scripts/triggers-end-to-end-tests/run.spec.js

```diff
@@ -0,0 +1,306 @@
+
+  this.rtdb_from_rtdb = false;
+  this.firestore_from_firestore = false;
+
+  this.emulators_process = null;
+}
+
+/*
+ * Check that all directions of database <-> functions <-> firestore
+ * worked.
+ */
+TriggerEndToEndTest.prototype.success = function success()
+{
+  return this.rtdb_from_firestore &&
+      this.rtdb_from_rtdb &&
+      this.firestore_from_firestore &&
+      this.firestore_from_rtdb;
+};
+
+TriggerEndToEndTest.prototype.startEmulators = function startEmulators()
+{
+  var self = this;
+  self.emulators_process = subprocess.spawn("node", [
+      PROJECT_ROOT + "/lib/bin/firebase.js",
+      "emulators:start",
+      "--project",
+      FIREBASE_PROJECT
+  ]);
+
+  self.emulators_process.stdout.on("data", function (data) {
+    process.stdout.write("[emulators stdout] " + data);
+    if (data.indexOf(RTDB_FUNCTION_LOG) > -1) {
+      self.rtdb_trigger_count++;
+    }
+    if (data.indexOf(FIRESTORE_FUNCTION_LOG) > -1) {
+      self.firestore_trigger_count++;
+    }
+  });
+
+  self.emulators_process.stderr.on("data", function (data) {
+    console.log("[emulators stderr] " + data);
+  });
+}
+
+TriggerEndToEndTest.prototype.stopEmulators = function stopEmulators(done)
+{
+  this.emulators_process.once('close', function (exitCode, signal) {
+    done();
+  });
+
+  /*
+   * CLI process only shuts down emulators cleanly on SIGINT.
+   */
+  this.emulators_process.kill("SIGINT");
+}
+
+TriggerEndToEndTest.prototype.writeToRtdb = function writeToRtdb(done)
+{
+  var url = "http://localhost:" + [
+      this.functions_emulator_port,
+      FIREBASE_PROJECT,
+      FIREBASE_PROJECT_ZONE,
+      "writeToRtdb"
+  ].join("/");
+
+  const req = request.get(url);
+
+  req.once('response', function(response) {
+    done(null, response);
+  });
+
+  req.once("error", function (err) {
+    done(err);
+  });
+};
+
+TriggerEndToEndTest.prototype.writeToFirestore = function writeToFirestore(done)
+{
+  var url = "http://localhost:" + [
+      this.functions_emulator_port,
+      FIREBASE_PROJECT,
+      FIREBASE_PROJECT_ZONE,
+      "writeToFirestore"
+  ].join("/");
+
+  const req = request.get(url);
+
+  req.once('response', function(response) {
+    done(null, response);
+  });
+  req.once('error', function (err) {
+    done(err);
+  });
+};
+
+function readConfig(done)
+{
+  fs.readFile("firebase.json", function (err, data) {
+    if (err) {
+      done(err);
+      return;
+    }
+    var config;
+    try {
+      config = JSON.parse(data);
+    } catch (err) {
+      done(err);
+      return;
+    }
+    done(null, config);
+  });
+}
+
+describe("database and firestore emulator function triggers", function () {
+  var test;
+
+  before(function (done) {
+    readConfig(function (err, config) {
+      if (err) {
+        done(new Error("error reading firebase.json: " + err));
+        return;
+      }
+      test = new TriggerEndToEndTest(config);
+      done();
+    });
+  });
+
+  before(function (done) {
```

Comment:
I would recommend pushing these two `before` blocks into one since I don't think there is a guarantee on calling order (besides that it's just convention to have a single before and a single after)

Reply:
Oh now that I am reading I see that there are four `before()` blocks ... can you explain what you were trying to do?

Reply:
I think the blocks are called in the order they were defined: https://mochajs.org/#run-cycle-overview. I was using this assumption to isolate different steps of the setup process.

Reply:
I have merged the setup code into a single before(...) and chained the individual setup steps together with an `async.series` call.